### PR TITLE
core(persistence): fix resource leaks - force closing files

### LINKED ISSUE #16823: FileStorage does not handle properly an empty yml file
Under OpenCV 4.2.*, using cv::FileStorage to open an (invalid) empty yml file (0 bytes) results in resource leaking by not closing files properly.
With current 4.3.0 dev, there is even a crash by nullptr dereference when trying to check the BOM.

---

relates #16488

### PR DESCRIPTION
**Pull request by alalek**
resolves #16823
closes #16822

### CODE REVIEW

File: modules/core/test/test_io.cpp [hunk 1]
```diff
@@ -1711,4 +1711,34 @@ TEST(Core_InputOutput, FileStorage_JSON_VeryLongLines)
     }
 }
 
+TEST(Core_InputOutput, FileStorage_empty_16823)
+{
+    std::string fname = "test_empty.yml";
```
Comment (mshabunin) [inline modules/core/test/test_io.cpp]:
```suggestion
    std::string fname = tempfile("test_empty.yml");
```

File: modules/core/test/test_io.cpp [hunk 2]
```diff
@@ -1711,4 +1711,34 @@ TEST(Core_InputOutput, FileStorage_JSON_VeryLongLines)
...
+    {
+        // create empty file
+        std::ofstream f(fname.c_str(), std::ios::out);
+    }
+
+    try
+    {
+        FileStorage fs(fname, FileStorage::READ);
+        ADD_FAILURE() << "Failed: Exception throws on empty input. Actual it doesn't.";
```
Comment (mshabunin) [inline modules/core/test/test_io.cpp]:
```suggestion
        ADD_FAILURE() << "Exception must be thrown for empty input.";
```

**Comment by mshabunin** [comment]
Maybe it will be even better to use `EXPECT_THROW(..., cv::Exception)`?

**Comment by alalek** [comment]
@chacha21 updated with handling of proposed use case

**Pull request merged**